small/obuf_asan.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "obuf.h"
+#include "util.h"
+
+#include <string.h>
+
+void
+obuf_create(struct obuf *buf, struct slab_cache *slabc, size_t start_capacity)
+{
+	buf->slabc = slabc;
+	buf->start_capacity = start_capacity;
+	size_t size = (SMALL_OBUF_IOV_MAX + 1) * sizeof(*buf->iov);
+	buf->iov = small_xmalloc(size);
+	memset(buf->iov, 0, size);
+	memset(buf->capacity, 0,
+	       lengthof(buf->capacity) * sizeof(buf->capacity[0]));
+	buf->pos = 0;
+	buf->used = 0;
+	buf->reserved = 0;
+}
+
+/** Allocate new memory block whether for allocation or reservation. */
+static void *
+obuf_prepare_buf(struct obuf *buf, size_t size)
+{
+	if (buf->pos >= SMALL_OBUF_IOV_CHECKED_SIZE - 1) {
+		int gpos = buf->pos - SMALL_OBUF_IOV_CHECKED_SIZE;
+		if (gpos < 0 ||
+		    (buf->iov[buf->pos].iov_len + size) > buf->capacity[gpos]) {
+			size_t capacity = buf->start_capacity << (gpos + 1);
+			while (capacity < size)
+				capacity <<= 1;
+			buf->pos++;
+			small_asan_assert(buf->pos < SMALL_OBUF_IOV_MAX);
+			struct iovec *iov = &buf->iov[buf->pos];
+			iov->iov_base = small_xmalloc(capacity);
```

Comment:
Why don't you use `small_asan_alloc` here as well?

Reply:
We only use it in `obuf` for alignment. It does not make much sense to align when we reach geometric part.